### PR TITLE
Read n zebra facts at a time

### PR DESCRIPTION
**Pull request by tranma**
We will read some maximum number of zebra facts per entity before stopping and calling compute. I think this is a good place to check the allocation limit (will be separate PR).

! @jystic @amosr 
/jury approved @amosr

### CODE REVIEW

File: data/sea/54-zebra.h [hunk 1]
```diff
@@ -3,6 +3,9 @@
 #if !ICICLE_NO_INPUT
 #if ICICLE_ZEBRA
 
+/* Number of facts to read before starting compute and checking allocation limit */
+static const iint_t zebra_chunk_fact_count = 2;
```
Comment (amosr) [inline data/sea/54-zebra.h]: is it worth putting this in `zebra_config_t` to make it easier to quickcheck with different limits?

File: data/sea/54-zebra.h [hunk 2]
```diff
@@ -79,6 +103,7 @@ static ierror_msg_t zebra_translate
     /* temporary fix until icicle uses the same time as zebra */
     itime_t zebra_chord_time = itime_to_epoch_seconds (chord_time);
 
+    /* the number of facts before chord time for this entity */
     for (iint_t i = 0; i != row_count; ++i) {
```
Comment (amosr) [inline data/sea/54-zebra.h]: this is counting all the facts for every call, but could you just loop from `state->entity_fact_offset` up to `state->entity_fact_offset + zebra_chunk_fact_count`?

File: data/sea/54-zebra.h [hunk 3]
```diff
@@ -87,25 +112,38 @@ static ierror_msg_t zebra_translate
         fact_count++;
     }
 
+    /* the number of unread facts for this entity */
+    iint_t fact_offset    = state->entity_fact_offset;
+    iint_t fact_remaining = fact_count == 0 ? 0 : fact_count - fact_offset;
```
Comment (amosr) [inline data/sea/54-zebra.h]: I would have thought `fact_count = 0 ==> state->entity_fact_offset = 0` anyway

File: data/sea/54-zebra.h [hunk 4]
```diff
@@ -87,25 +112,38 @@ static ierror_msg_t zebra_translate
...
+        fact_count = fact_remaining;
+        state->entity_fact_offset = 0;
+    }
+
+    state->fact_count += fact_count;
+
     itime_t *sadface_times = anemone_mempool_alloc (mempool, fact_count * sizeof (itime_t));
 
     for (iint_t i = 0; i != fact_count; ++i) {
         sadface_times[i] = itime_from_epoch_seconds (attribute->times[i]);
```
Comment (amosr) [inline data/sea/54-zebra.h]: `attribute->times[i + fact_offset]`

File: data/sea/54-zebra.h [hunk 5]
```diff
@@ -87,25 +112,38 @@ static ierror_msg_t zebra_translate
...
     }
 
-    *new_count = fact_count;
-    *new_tombstone = (ierror_t*)attribute->tombstones;
-    *new_fact_time = sadface_times;
-
-    zebra_translate_table (mempool, 0, fact_count, dst, &attribute->table);
+    *new_fact_count = fact_count;
+    *new_tombstone  = (ierror_t*) attribute->tombstones;
```
Comment (amosr) [inline data/sea/54-zebra.h]: `attribute->tombstones + fact_offset`, I think

File: data/sea/54-zebra.h [hunk 6]
```diff
@@ -116,14 +154,15 @@ static int64_t zebra_translate_table
...
-        offset += zebra_translate_column (mempool, elem_start, count, dst + offset, cols + i);
+        offset += zebra_translate_column (mempool, fact_offset, elem_start, count, dst + offset, cols + i);
     }
 
     return offset;
 }
 
 static int64_t zebra_translate_column
     ( anemone_mempool_t *mempool
+    , iint_t fact_offset
```
Comment (amosr) [inline data/sea/54-zebra.h]: looks like you should be able to just use `elem_start` instead of adding this new parameter

File: data/sea/54-zebra.h [hunk 7]
```diff
@@ -156,7 +195,7 @@ static int64_t zebra_translate_column
                 int offset = 1;
                 for (int i = 0; i != count; ++i) {
                     int len = data.a.n[i];
-                    offset = zebra_translate_table (mempool, elem_start, len, (void**)(strings + i), &data.a.table);
+                    offset = zebra_translate_table (mempool, fact_offset, elem_start, len, (void**)(strings + i), &data.a.table);
```
Comment (amosr) [inline data/sea/54-zebra.h]: this was already half-wrong - it didn't support nested arrays properly which wasn't a problem, but now we need to slice out of the nested arrays

so you need to loop through from `[0,elem_start)` summing up the lengths and use that as `elem_start` for the recursive calls.
but this loop wouldn't be necessary after #473 since latest zebra stores them as prefix sum
anyway.

we might need another function to sum the offsets of a table without doing any translation - because at the moment if the array is empty the offset isn't being computed properly and just set to 1. that's not directly related to this PR but it's probably worth working this out at the same time if you want to go through it together tomorrow

File: data/sea/54-zebra.h [hunk 8]
```diff
@@ -204,22 +246,38 @@ void zebra_collect_state (zebra_config_t *cfg, zebra_state_t *state)
...
+
+    ierror_loc_t  error_loc = psv_configure_fleet ((char*) entity->id_bytes, entity->id_length, piano, state->fleet);
+    if (error_loc) return ierror_loc_pretty (error_loc, 0);
+
+    ierror_msg_t error = zebra_read_entity (piano, state, entity);
+    if (error) return error;
+
+    /* keep going if we have not finished reading all facts for this entity */
+    if (state->entity_fact_offset != 0) {
+        zebra_read_step (piano, state, entity);
```
Comment (amosr) [inline data/sea/54-zebra.h]: this should return the result in case it's an error

File: data/sea/54-zebra.h (hunk 6 above)
Comment (tranma) [inline data/sea/54-zebra.h]: yeah - I had that originally, but thought this is clearer. You think I should just roll it in to `elem_start`?

File: data/sea/54-zebra.h (hunk 3 above)
Comment (tranma) [inline data/sea/54-zebra.h]: yeah, this was wrong -- `fact_count == 0` should only be when we have not read any fact for this entity. That should be an invariant actually.

File: data/sea/54-zebra.h (hunk 1 above)
Comment (tranma) [inline data/sea/54-zebra.h]: sure, we have this in the config for psv too.

File: data/sea/54-zebra.h (hunk 6 above)
Comment (amosr) [inline data/sea/54-zebra.h]: I think it's cleaner to just have one, that way it's just performing a slice from `[elem_start,elem_start+fact_count)` or something

File: data/sea/54-zebra.h [hunk 9]
```diff
@@ -68,44 +88,72 @@ static ierror_msg_t zebra_translate
...
+        /* the number of facts before chord time for this entity */
+        for (iint_t i = 0; i != row_count; ++i) {
+            itime_t time = attribute->times[i];
+            if (zebra_chord_time <= time) {
+                break;
+            }
+            fact_count++;
         }
-        fact_count++;
+        state->entity_fact_count = fact_count;
```
Comment (jystic) [inline data/sea/54-zebra.h]: so this just runs the first time in here for each entity?

Comment (tranma) [inline data/sea/54-zebra.h]: yeah

File: data/sea/54-zebra.h [hunk 10]
```diff
@@ -68,44 +88,72 @@ static ierror_msg_t zebra_translate
...
 
+    /* number of unread facts for this entity */
+    iint_t fact_offset    = state->entity_fact_offset;
+    iint_t fact_remaining = fact_count - fact_offset;
+
+    if (fact_remaining > zebra_chunk_fact_count) {
+        fact_count = zebra_chunk_fact_count;
+        state->entity_fact_offset += fact_count;
+    } else {
+        fact_count = fact_remaining;
```
Comment (jystic) [inline data/sea/54-zebra.h]: I don't like how this variable is written to so many times throughout the function, makes it a bit hard to follow what the meaning of it is.

Comment (jystic) [inline data/sea/54-zebra.h]: as in maybe it could get a fresh name at this point, rather than re-using the `fact_count` you were incrementing above, like `fact_chunk_count` or something? /not sure

Comment (tranma) [inline data/sea/54-zebra.h]: yeah fair enough

Comment (jystic) [inline data/sea/54-zebra.h]: i just think we should try to write our C like Haskell where possible, pretend it doesn't have mutable variables where possible

**Comment by tranma** [comment]
this is ok but I'll just leave it until there are "peek input" tests for psv and zebra.

File: data/sea/54-zebra.h [hunk 11]
```diff
@@ -206,22 +270,42 @@ void zebra_collect_state (zebra_config_t *cfg, zebra_state_t *state)
...
+    ierror_loc_t  error_loc = psv_configure_fleet ((char*) entity->id_bytes, entity->id_length, piano, state->fleet);
+    if (error_loc) return ierror_loc_pretty (error_loc, 0);
+
+    ierror_msg_t error = zebra_read_entity (piano, state, entity);
+    if (error) return error;
+
+    /* keep going if we have not finished reading all facts for this entity */
+    for (int64_t i = 0; i != state->attribute_count; ++i) {
+        if (state->entity_fact_offset[i] != state->entity_fact_count[i]) {
+            return zebra_read_step (piano, state, entity);
```
Comment (amosr) [inline data/sea/54-zebra.h]: I think this would be clearer as a loop

File: data/sea/54-zebra.h [hunk 12]
```diff
@@ -68,38 +84,63 @@ static ierror_msg_t zebra_translate
...
+        sadface_times[i] = itime_from_epoch_seconds (attribute->times[i + fact_offset]);
+    }
 
-    zebra_translate_table (mempool, 0, fact_count, dst, &attribute->table);
+    *new_fact_count = fact_to_read;
+    *new_tombstone  = (ierror_t*) attribute->tombstones + fact_offset;
+    *new_fact_time  = sadface_times;
 
-    state->fact_count += fact_count;
+    zebra_translate_table (mempool, fact_offset, fact_to_read, dst, &attribute->table);
```
Comment (amosr) [inline data/sea/54-zebra.h]: this function looks right

File: data/sea/54-zebra.h [hunk 13]
```diff
@@ -152,12 +209,13 @@ static int64_t zebra_translate_column
...
                 istring_t *strings = anemone_mempool_alloc (mempool, count * sizeof (istring_t));
-                int offset = 1;
                 for (int i = 0; i != count; ++i) {
-                    int len = data.a.n[i];
-                    offset = zebra_translate_table (mempool, elem_start, len, (void**)(strings + i), &data.a.table);
-                    elem_start += len;
+                    int64_t index = i + elem_start;
+                    int64_t start = data.a.n[index]   - data.a.n[0];
+                    int64_t len   = data.a.n[index+1] - data.a.n[index];
+                    offset = zebra_translate_array (mempool, start, len, (void**)(strings + i), &data);
```
Comment (amosr) [inline data/sea/54-zebra.h]: this function looks right

Comment (amosr) [inline data/sea/54-zebra.h]: can we add an assertion that

```
data.a.table.columns_count == 1 &&
data.a.table.columns[0].type == ZEBRA_BYTE
```

and if it's false just die really badly saying we don't support that input format

File: data/sea/54-zebra.h [hunk 14]
```diff
@@ -206,22 +270,42 @@ void zebra_collect_state (zebra_config_t *cfg, zebra_state_t *state)
     free (state);
```
Comment (amosr) [inline data/sea/54-zebra.h]: don't need to free the buffer, entity_fact_offset etc? (I don't think this is important)

File: data/sea/54-zebra.h (hunk 13 above)
Comment (tranma) [inline data/sea/54-zebra.h]: yeah, that sounds good

File: data/sea/54-zebra.h [hunk 15]
```diff
@@ -206,22 +270,42 @@ void zebra_collect_state (zebra_config_t *cfg, zebra_state_t *state)
...
+    /* keep going if we have not finished reading all facts for this entity */
+    for (int64_t i = 0; i != state->attribute_count; ++i) {
+        if (state->entity_fact_offset[i] != state->entity_fact_count[i]) {
+            return zebra_read_step (piano, state, entity);
+        }
+        state->entity_fact_count [i] = 0;
+        state->entity_fact_offset[i] = 0;
+    }
+
+    return 0;
```
Comment (amosr) [inline data/sea/54-zebra.h]: yeah this is weird control flow, but I think something like this would be cleaner:

```
while (true) {
  read_entity
  all = true;
  for (...) if (fact_offset != fact_count) all = false;
  if (!all) break;
}

for (...) fact_offset = fact_count = 0;
```

Comment (amosr) [inline data/sea/54-zebra.h]: you've hand fused the two loops!

**Comment by jystic** [comment]
How's this one going?

**Comment by tranma** [comment]
@jystic it's ok on its own but I'd like to finish & merge the zebra quickcheck tests in `topic/input-test` first

File: data/sea/54-zebra.h (hunk 15 above)
Comment (tranma) [inline data/sea/54-zebra.h]: haha I guess it was a bit too haskelly. I made it 2 loops

**Comment by tranma** [comment]
this should be ok now with the tests in

**Comment by amosr** [comment]
👍
